source/Halibut/ServiceModel/HalibutProxy.cs

```diff
@@ -165,9 +153,39 @@ static void EnsureNotError(ResponseMessage responseMessage)
             if (responseMessage.Error == null)
                 return;
 
-            var realException = responseMessage.Error.Details as string;
-            throw new HalibutClientException(responseMessage.Error.Message, realException);
+            ThrowExceptionFromReceivedError(responseMessage.Error);
+        }
+        
+        internal static void ThrowExceptionFromReceivedError(ServerError error)
+        {
+            var realException = error.Details as string;
+            if (!string.IsNullOrEmpty(error.HalibutErrorType))
+            {
+                var theType = Type.GetType(error.HalibutErrorType);
```

Comment:
Should we wrap this in a try-catch so we guarantee we don't break things unexpectedly and instead fallback to the previous behaviour?

Reply:
If a non HalibutClientException is thrown, that error is now caught and logged. Finally the old behaviour is run of just throwing a HalibutClientException.